File: android/app/src/main/java/com/obubba/app/plugins/WidgetBridgePlugin.kt
package com.obubba.app.plugins

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.SharedPreferences
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.obubba.app.widgets.OBubbaSummaryWidget

/**
 * Bridges widget data between the web app and Android widgets.
 * Writes data to SharedPreferences so widgets can read it.
 */
@CapacitorPlugin(name = "OBWidgetBridge")
class WidgetBridgePlugin : Plugin() {

    private fun getSharedPrefs(): SharedPreferences {
        return context.getSharedPreferences("obubba_widget_data", Context.MODE_PRIVATE)
    }

    @PluginMethod
    fun setData(call: PluginCall) {
        val json = call.getString("json") ?: run {
            call.reject("json is required")
            return
        }

        getSharedPrefs().edit().putString("widgetData", json).apply()

        // Trigger widget update
        val appWidgetManager = AppWidgetManager.getInstance(context)
        val widgetComponent = ComponentName(context, OBubbaSummaryWidget::class.java)
        val widgetIds = appWidgetManager.getAppWidgetIds(widgetComponent)
        if (widgetIds.isNotEmpty()) {
            appWidgetManager.notifyAppWidgetViewDataChanged(widgetIds, android.R.id.list_container)
            OBubbaSummaryWidget.updateWidgets(context, appWidgetManager, widgetIds)
        }

        val ret = com.getcapacitor.JSObject()
        ret.put("saved", true)
        call.resolve(ret)
    }

    @PluginMethod
    fun reloadAll(call: PluginCall) {
        val appWidgetManager = AppWidgetManager.getInstance(context)
        val widgetComponent = ComponentName(context, OBubbaSummaryWidget::class.java)
        val widgetIds = appWidgetManager.getAppWidgetIds(widgetComponent)
        if (widgetIds.isNotEmpty()) {
            OBubbaSummaryWidget.updateWidgets(context, appWidgetManager, widgetIds)
        }
        call.resolve()
    }
}

File: ios/App/App/Plugins/WidgetBridgePlugin.swift
import Foundation
import Capacitor
import WidgetKit

/// Bridges widget data between the web app and iOS WidgetKit.
/// Writes data to a shared App Group container file so widgets can read it.
/// Uses file-based sharing instead of UserDefaults to avoid CFPrefs container issues.
@objc(OBWidgetBridge)
public class WidgetBridgePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "OBWidgetBridge"
    public let jsName = "OBWidgetBridge"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setData", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reloadAll", returnType: CAPPluginReturnPromise),
    ]

    private let appGroupId = "group.com.obubba.app"

    /// Returns the shared file URL for widget data
    private func sharedFileURL() -> URL? {
        FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupId)?
            .appendingPathComponent("widgetData.json")
    }

    @objc func setData(_ call: CAPPluginCall) {
        guard let json = call.getString("json") else {
            call.reject("json is required")
            return
        }

        // Method 1: Write to shared file (primary — most reliable)
        if let fileURL = sharedFileURL() {
            do {
                try json.write(to: fileURL, atomically: true, encoding: .utf8)
                print("[OBWidgetBridge] Wrote \(json.count) bytes to \(fileURL.path)")
            } catch {
                print("[OBWidgetBridge] File write error: \(error)")
            }
        } else {
            print("[OBWidgetBridge] ERROR: Could not get App Group container URL")
        }

        // Method 2: Also write to UserDefaults (backup)
        if let defaults = UserDefaults(suiteName: appGroupId) {
            defaults.set(json, forKey: "widgetData")
            defaults.synchronize()
        }

        // Tell WidgetKit to refresh
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }

        call.resolve(["saved": true])
    }

    @objc func reloadAll(_ call: CAPPluginCall) {
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }
        call.resolve()
    }
}
